src/main/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -317,9 +317,15 @@ public SamReader open(final SamInputResource resource) {
                             // do not close bufferedStream, it's the same stream we're getting here.
                             SeekableStream sourceSeekable = data.asUnbufferedSeekableStream();
                             if (indexFile!=null || null == sourceSeekable || null == indexSeekable) {
-                                // not seekable.
-                                // it's OK that we consumed a bit of the stream already, this ctor expects it.
-                                primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                                if (null == sourceSeekable || null == indexSeekable) {
+                                    // not seekable.
+                                    // it's OK that we consumed a bit of the stream already, this ctor expects it.
+                                    primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                                } else {
+                                    sourceSeekable.seek(0);
+                                    primitiveSamReader = new BAMFileReader(
+                                            sourceSeekable, indexSeekable, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                                }
                             } else {
```

Comment:
I think that this entire nested if statement from line 319 up to line 336 can be simplified from this:

```
                            if (indexFile!=null || null == sourceSeekable || null == indexSeekable) {
                                if (null == sourceSeekable || null == indexSeekable) {
                                    // not seekable.
                                    // it's OK that we consumed a bit of the stream already, this ctor expects it.
                                    primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
                                } else {
                                    sourceSeekable.seek(0);
                                    primitiveSamReader = new BAMFileReader(
                                            sourceSeekable, indexSeekable, false, asynchronousIO, validationStringency, this.samRecordFactory);
                                }
                            } else {
                                // seekable.
                                // need to return to the beginning because it's the same stream we used earlier
                                // and read a bit from, and that form of the ctor expects the stream to start at 0.
                                sourceSeekable.seek(0);
                                primitiveSamReader = new BAMFileReader(
                                    sourceSeekable, indexSeekable, false, asynchronousIO, validationStringency, this.samRecordFactory);
                            }
```

to this:

```
                                if (null == sourceSeekable || null == indexSeekable) {
                                    // not seekable.
                                    // it's OK that we consumed a bit of the stream already, this ctor expects it.
                                    primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
                                } else {
                                    // seekable.
                                    // need to return to the beginning because it's the same stream we used earlier
                                    // and read a bit from, and that form of the ctor expects the stream to start at 0.
                                    sourceSeekable.seek(0);
                                    primitiveSamReader = new BAMFileReader(
                                            sourceSeekable, indexSeekable, false, asynchronousIO, validationStringency, this.samRecordFactory);
                                }
```

Since the two `else` clauses in the original code are identical.

In other words, what you seem to have done with this patch is just to say "if you have an index that's a file, you can now use the constructor that takes two seekable streams instead of the (non-seekable-stream, file) constructor."